### DISCUSSION TITLE
[help] Extreme memory load from `tar_quarto_rep()` on an qmd with `tar_load(df, branches = params$branch)`

### DISCUSSION THREAD

**Discussion by asadow** [ropensci/targets#1178] [Help] [answered]
### Help

- [X] I understand and agree to <https://books.ropensci.org/targets/help.html>.

### Description

In my situation, `df` is a pretty big target (1000 ggplots) that has been branched off. Without branching it would be ~64MB.

I've got a decent system; I was able to create `df` using distributed computing very fast (~15 secs). So here is where I'm surprised and I can't figure out what's going on.

I am creating separate renders for each of the branches. A single worker is taking up 10GB of memory. So I can't use multiple workers. I've tried all the options in the Performance chapter. Am I missing something?

**Comment by wlandau** [comment]
It would help to look at a small reprex.

**Reply by snowpong** [reply]
Possibly your ggplots are huge? When saved they can store the complete parent environment as part of themselves. It's (somewhat) possible to strip ggplots of this. See https://stackoverflow.com/a/77373906/1666063 for something to try.

**Reply by asadow** [reply]
Thank you both. A reprex is now available at https://github.com/asadow/targets_calendars. I was hopeful that stripping the ggplots of the parent environment would fix it but alas.

**Reply by wlandau** [reply]
Had a look at the reprex, and it seems to reference a `Rmd/calendar_child.Rmd` document. Do you have that one?

**Reply by wlandau** [reply]
But the memory problem is that `tarchetypes` looks at `tar_load(calendars, ...)` and thinks it needs to load all branches. `tar_render_rep()` creates this command in the pipeline:

```r
tarchetypes::tar_render_rep_run(path = "/CENSORED/targets_calendars/Rmd/calendar.Rmd", params = reports_calendars_params, args = list("/CENSORED/targets_calendars/Rmd/calendar.Rmd", getwd(), TRUE), deps = list(calendars), rep_workers = 1L)
```

The `deps = list(calendars)` is an artifact of the static code analysis and there is no way to make it aware that it needs a specific dynamic branch.

The solution is to write `retrieval = "none"` in `tar_render_rep()`. In your case, you want `tar_load()` to take care of all the loading anyway, so you don't need `targets` to load anything before running the report.

**Reply by asadow** [reply]
Thank you Will. I tried your suggestion but then it looks like the `params` argument causes me trouble. I get the error `object active_employees is not found` despite it being a up-to-date target. 

I added the missing `Rmd/calendar_child.Rmd` to the repo.

**Reply by wlandau** [reply]
In that case, it gets tricky. You may have to save each branch of `employees` as a `format = "file"` target and then have each Quarto report manually read the contents. There are other ways, but they seem messier.

**Reply by asadow** [reply]
Do you have any examples I can refer to for that?

Even without the reliance on `active_employees` by setting `params = 1:300`, I get `Last error: object 'reports_calendars_params' not found`.

**Reply by wlandau** [reply]
On second thought: `tar_quarto_rep()` does not fit this scenario because it only branches over Quarto parameters. All other dependencies are either non-dynamic or an aggregate of dynamic branches. What you are really trying to do is branch over `calendars`, and that should be done with an explicit `tar_target(name, command = quarto::quarto_render("calendars.Rmd", ...), pattern = map(calendars))`.

**Reply by asadow** [reply]
Ok, great. I'm stuck however on how to name the files based on the employee. Without `output_file`, a single `calendar.html` is made and is rewritten over. But the below code will give `Object active_employees not found`.

```
tar_target(
    reports_calendars, 
    command = quarto::quarto_render(
      here::here("calendar.Rmd"),
      output_file = paste0(
      1:length(active_employee), ".html"
      )
      ), 
    pattern = map(calendars)
    )
```

**Reply by wlandau** [reply]
You would need something from the current branch of `calendars`. When you map over `calendars`, the mention of `calendars` inside your command is the current branch. If `calendars` is a data frame with a column to indicate the output file, it would look something like this:

```r
tar_target(
  reports_calendars, 
  command = quarto::quarto_render(
    here::here("calendar.Rmd"),
    output_file = calendars$output_file[1]
  ), 
  pattern = map(calendars)
)
```

Each branches `output_file` should be different because `calendars$output_file[1]` uses only the subset of `calendars` corresponding to the current branch of `reports_calendars`.

**Reply by asadow** [reply]
Ah perfect. But how would you load in the branch inside the Rmd? I feel like I need a way to refer to the pattern, like `tar_load(calendars, branches = pattern)`.

I've tried `tar_load(calendars)` with and without params$branch, and omitting `tar_load()` completely.

**Reply by wlandau** [reply, marked as answer]
If you really need to use Quarto, then you would have to make a dynamic file target to save each section of `calendars` to a file, then map over those files in `report_calendars`. But if you can switch to R Markdown, the report is rendered in the same process that runs the target, so then you could just reference the object in the R code of the report. Either way, `tar_load()` and `tar_read()` are not meant for this scenario.

**Reply by asadow** [reply]
Thank you dearly. As you implied, it works when I omit `tar_load()` from the report and use `rmarkdown::render()` instead of `quarto::quarto_render()` in `_targets.R`. The issue was in using Quarto.

I tend to avoid Quarto in tandem with targets. For one, I always get `quarto.exe failed` when the Rmd/qmd file is not in the root directory. And this adds another reason.

**Reply by asadow** [reply]
@wlandau I spoke too soon: It works for a handful of reports, but suddenly one report errors with `Last error: pandoc document conversion failed with error 1`.

[Repo reprex here.](https://github.com/asadow/targets_calendars)

I can't reproduce the error; manually running reports works fine. I tried restarting and running `targets::tar_make(callr_function = NULL)`.

**Reply by wlandau** [reply]
I was able to reproduce the error you saw, and it indeed does go away when I render the report locally or from within the [built-in `targets` interactive debugging system](https://books.ropensci.org/targets/debugging.html#pause-the-pipeline-with-the-debug-option). Then I tried `tar_make(use_crew = FALSE)`, and everything worked.

So I think the pandoc error happens because two parallel processes from `crew` are trying to write the same intermediate files at the same time. Even though your HTML output files are all unique, `pandoc` may still write temporary local intermediate files based on the name of the common Rmd file. If the `crew` workers fight over those intermediate files, the results are unpredictable.

So for each `rmarkdown::render()` call, I recommend making a copy of `Rmd/calendar.Rmd` to a temporary file (c.f. `tempfile()` and `tempdir()`), rendering it from the temporary location, and then moving the HTML file to the output folder you want.

As a last resort, you could alternatively set `deployment = "main"` in the report targets so they don't run on `crew` workers.

**Reply by asadow** [reply]
I did what you recommended and it works fantastically. Thanks Will!

**Reply by wlandau** [reply]
Glad to hear it works now.
